### PR TITLE
Support formatting and mounting GCE PD without 'safe_format_and_mount'

### PR DESCRIPTION
**Pull request by ghost**
The GCE PD plugin uses safe_format_and_mount found on standard GCE images:

https://github.com/GoogleCloudPlatform/compute-image-packages/blob/master/google-startup-scripts/usr/share/google/safe_format_and_mount

On custom images where this is not available pods fail to format and mount GCE PDs.

This patch uses linux utilities in a similar way to the safe_format_and_mount
script to format and mount the GCE PD device. That is first attempt a mount.
If mount fails and detects a file system "null" try to use file to investigate
the device. If 'file' fails to get any information about the device and simply
returns "data" then assume the device is not formatted and format it and attempt
to mount it again.

### CODE REVIEW

**Comment by swagiaal** [comment]
@ddysher What do you think ?
@childsb with this PR plus #7847 GCE PD works out of the box on Atomic

File: pkg/volume/gce_pd/gce_util.go [hunk 1]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
-	cmd := mounter.runner.Command("/usr/share/google/safe_format_and_mount", args...)
+	safeFormatAndMountPath := "/usr/share/google/safe_format_and_mount"
+	glog.V(5).Infof("exec-ing: %s %v", safeFormatAndMountPath, args)
+	cmd := mounter.runner.Command(safeFormatAndMountPath, args...)
 	dataOut, err := cmd.CombinedOutput()
 	if err != nil {
-		glog.V(5).Infof("error running /usr/share/google/safe_format_and_mount\n%s", string(dataOut))
+		glog.V(5).Infof("error running %s\n%s", safeFormatAndMountPath, string(dataOut))
+		// if safe_format_and_mount is not available on this system use linux utilities directly.
+		if strings.Contains(err.Error(), safeFormatAndMountPath+": no such file or directory") {
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: This is brittle, at least, you need to lowercase err.Error().  BTW, I think shell use upper case for error message?, i.e. No such file or directory.

Also, why not check file existence first?

File: pkg/volume/gce_pd/gce_util.go [hunk 2]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+			return mounter.formatAndMount(source, target, fstype, options)
+		}
+	}
+	return err
+}
+
+// Uses unix utils to format and mount the given disk
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: godoc generall starts with func name, so change this to:
`formatAndMount uses unix utils to format and mount the given disk.`

File: pkg/volume/gce_pd/gce_util.go [hunk 3]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+func (mounter *gceSafeFormatAndMount) formatAndMount(source string, target string, fstype string, options []string) error {
+
+	// Add the default options from safe_format_and_mount
+	options = append(options, "discard")
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: Why do you want TRIM here?

File: pkg/volume/gce_pd/gce_util.go [hunk 4]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+	options = append(options, "defaults")
+
+	// Try to mount the disk. Let 'mount' try to determine the fstype.
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: Is there a reason to ignore 'fstype'?

File: pkg/volume/gce_pd/gce_util.go [hunk 5]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+	err := mounter.Interface.Mount(source, target, "", options)
+	if err != nil && strings.Contains(err.Error(), "unknown filesystem type '(null)'") {
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: Again, why not call diskLooksUnformatted first?

File: pkg/volume/gce_pd/gce_util.go [hunk 6]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+			if err == nil {
+				// the disk has been formatted sucessfully try to mount it again.
+				err = mounter.Interface.Mount(source, target, fstype, options)
+			}
+		}
 	}
 	return err
 }
+
+// uses 'file' to see if the given disk is unformated
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: Same here, godoc

File: pkg/volume/gce_pd/gce_util.go (hunk 3 above)
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: Looks like you just copied the script.

File: pkg/volume/gce_pd/gce_util.go (hunk 1 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: I have changed this to check for the existence of the file instead of trying to match the error.

File: pkg/volume/gce_pd/gce_util.go (hunk 2 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: done

File: pkg/volume/gce_pd/gce_util.go (hunk 3 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: Yeah I just copied the script here. I did a bit of research and I have not found great reviews of the "discard" option. I will remove it unless you say otherwise.

File: pkg/volume/gce_pd/gce_util.go (hunk 4 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: Yeah, if you let mount try to guess the fstype it will return "null" for an unformatted disk. If you supply the fstype mount returns "[...] wrong fs type [...]" meaning that the disk is either unformatted or you supplied the wrong filesystem type.
I have updated the comment to reflect that.

File: pkg/volume/gce_pd/gce_util.go (hunk 5 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: It is the combined failure of 'mount' with 'filesystem type (null)' and 'file' returning 'data' for the file that leads to the conclusion that disk is unformatted. If we were to call diskLooksUnformatted first we would still have to call mount to corroborate as 'file' can return 'data' for a file it does not recognized even if it is a formatted device. I don't think that is likely but that is what the man page states. the safe_format_and_mount script works the same way.
Also, in the life cycle of the disk it will only be formatted once but possibly mounted many times making the first mount call more likely to succeed and the 'diskLooksUnformatted' call unnecessary.

File: pkg/volume/gce_pd/gce_util.go (hunk 6 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: done

File: pkg/util/mount/mount_linux.go [hunk 7]
```diff
@@ -101,8 +102,9 @@ func doMount(source string, target string, fstype string, options []string) erro
 	if err != nil {
 		glog.Errorf("Mount failed: %v\nMounting arguments: %s %s %s %v\nOutput: %s\n",
 			err, source, target, fstype, options, string(output))
+		return errors.New(string(output))
```
Comment (pmorie) [inline pkg/util/mount/mount_linux.go]: +100, good catch

**Comment by swagiaal** [comment]
@ddysher I have addressed your comments and responded to your questions. What do you think ?

File: pkg/volume/gce_pd/gce_util.go [hunk 8]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
 	return err
 }
+
+// diskLooksUnformatted uses 'file' to see if the given disk is unformated
+func (mounter *gceSafeFormatAndMount) diskLooksUnformatted(disk string) (bool, error) {
+	args := []string{"-L", "--special-files", disk}
+	cmd := mounter.runner.Command("file", args...)
+	dataOut, err := cmd.CombinedOutput()
+
+	if err == nil && strings.TrimSpace(string(dataOut)) == disk+": data" {
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: https://github.com/GoogleCloudPlatform/compute-image-packages/blob/master/google-startup-scripts/usr/share/google/safe_format_and_mount#L90

The bash script uses *filesystem*, is it a better condition to check?

File: pkg/volume/gce_pd/gce_util.go [hunk 9]
```diff
@@ -149,11 +150,59 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+// formatAndMount uses unix utils to format and mount the given disk
+func (mounter *gceSafeFormatAndMount) formatAndMount(source string, target string, fstype string, options []string) error {
+
+	// Add the default options from safe_format_and_mount
+	options = append(options, "defaults")
+
+	// Try to mount the disk. Let 'mount' try to determine the fstype; if it determines the
+	// filesystem type to be "null" then assume the disk is unformatted.
+	err := mounter.Interface.Mount(source, target, "", options)
+	if err != nil && strings.Contains(err.Error(), "unknown filesystem type '(null)'") {
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: I'm not a fan of checking error message, but no better suggestion though.  Probably fine here :)

File: pkg/volume/gce_pd/gce_util_test.go [hunk 10]
```diff
@@ -21,13 +21,27 @@ import (
...
+}
+
+func (mounter ErrorMounter) Mount(source string, target string, fstype string, options []string) error {
+	if mounter.err != nil {
+		return mounter.err
+	}
+	return mounter.FakeMounter.Mount(source, target, fstype, options)
+}
+
 func TestSafeFormatAndMount(t *testing.T) {
```
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util_test.go]: Test "/usr/share/google/safe_format_and_mount" doesn't exist?

File: pkg/volume/gce_pd/gce_util.go (hunk 3 above)
Comment (ddysher) [inline pkg/volume/gce_pd/gce_util.go]: Yeah, I think it's generally suggested to disable the options by default; not sure why that script enables it.  We can add it back if people suggest otherwise.

**Comment by ddysher** [comment]
some minor comments.  make sure PD e2d passes.

File: pkg/volume/gce_pd/gce_util.go (hunk 8 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: Yeah that makes more sense. Done.

File: pkg/volume/gce_pd/gce_util.go (hunk 9 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: Yeah I could not think of another way either :/

File: pkg/volume/gce_pd/gce_util_test.go (hunk 10 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util_test.go]: Hmm.. I've thought about that. The problem is that it is difficult to do without being able to mock 'os'. It would be worthwhile to refactor persistent volumes so that the os object is injected and can be mocked for testing but that would have to be its own patch.

**Comment by swagiaal** [comment]
@ddysher I have updated the patch to address your comments. PD e2e tests are passing.

**Comment by ddysher** [comment]
LGTM
but need triage for code freeze

**Comment by thockin** [comment]
I want to have a look at this.

**Comment by rjnagal** [comment]
@thockin removing lgtm  while you take a look and decide.

**Comment by erictune** [comment]
@thockin push this to post-1.0?

**Comment by swagiaal** [comment]
@thockin @erictune it would be great of this can patch can make it into 1.0 as it enables GCE PD on Atomic Host and other custom images uploaded by users to GCE out of the box

**Comment by thockin** [comment]
to the goal of "custom images uploaded by users to GCE" there are a lot of things that you really want to do to in order to properly be a GCE image.

https://cloud.google.com/compute/docs/tutorials/building-images

File: pkg/volume/gce_pd/gce_util.go [hunk 11]
```diff
@@ -149,11 +150,60 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
-	glog.V(5).Infof("exec-ing: /usr/share/google/safe_format_and_mount %v", args)
-	cmd := mounter.runner.Command("/usr/share/google/safe_format_and_mount", args...)
+	safeFormatAndMountPath := "/usr/share/google/safe_format_and_mount"
+	glog.V(5).Infof("exec-ing: %s %v", safeFormatAndMountPath, args)
+	cmd := mounter.runner.Command(safeFormatAndMountPath, args...)
 	dataOut, err := cmd.CombinedOutput()
 	if err != nil {
-		glog.V(5).Infof("error running /usr/share/google/safe_format_and_mount\n%s", string(dataOut))
+		glog.V(5).Infof("error running %s\n%s", safeFormatAndMountPath, string(dataOut))
+		// if safe_format_and_mount is not available on this system use linux utilities directly.
```
Comment (thockin) [inline pkg/volume/gce_pd/gce_util.go]: Why support both ways?  If this is equivalent, then we should drop the external dependency.

File: pkg/volume/gce_pd/gce_util.go (hunk 9 above)
Comment (thockin) [inline pkg/volume/gce_pd/gce_util.go]: We could figure out what mount is actually doing under the covers?

File: pkg/volume/gce_pd/gce_util.go [hunk 12]
```diff
@@ -149,11 +150,60 @@ func (mounter *gceSafeFormatAndMount) Mount(source string, target string, fstype
...
+	// Add the default options from safe_format_and_mount
+	options = append(options, "defaults")
+
+	// Try to mount the disk. Let 'mount' try to determine the fstype; if it determines the
+	// filesystem type to be "null" then assume the disk is unformatted.
+	err := mounter.Interface.Mount(source, target, "", options)
+	if err != nil && strings.Contains(err.Error(), "unknown filesystem type '(null)'") {
+		// It is very likely that this disk is not formatted. Double check using 'file'
+		var notFormatted bool
+		notFormatted, err = mounter.diskLooksUnformatted(source)
```
Comment (thockin) [inline pkg/volume/gce_pd/gce_util.go]: Optional, maybe a TODO: Would be nice to check that the device is not partitioned.  e.g. if I say /dev/sdb and I meant sdb1, it should fail rather than blow away my data.

**Comment by thockin** [comment]
I am OK with this overall - it's pretty straightforward.  Why not get rid of support for the old legacy script and just do it ourselves always?

File: pkg/volume/gce_pd/gce_util.go (hunk 11 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: Done.

File: pkg/volume/gce_pd/gce_util.go (hunk 12 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_util.go]: I added a TODO there and will look into it in a separate patch.

**Comment by swagiaal** [comment]
@thockin I made the changes you asked for. Dropping the use of the script made writing the test easier so I expanded the test to better cover the code. I also removed the check for the specific error message as older versions of mount have a different message.
e2e PD tests are passing

**Comment by thockin** [comment]
The chaneg looks good, but I am nervous putting this in at this point.  I'd rather this wait until after 1.0 - is that going to derail anything?  

LGTM, hold until after 1.0

File: pkg/volume/aws_ebs/aws_util.go [hunk 13]
```diff
@@ -110,37 +107,3 @@ func (util *AWSDiskUtil) DetachDisk(pd *awsElasticBlockStore) error {
 	}
 	return nil
 }
-
-// safe_format_and_mount is a utility script on AWS VMs that probes a persistent disk, and if
-// necessary formats it before mounting it.
-// This eliminates the necessity to format a PD before it is used with a Pod on AWS.
-// TODO: port this script into Go and use it for all Linux platforms
```
Comment (swagiaal) [inline pkg/volume/aws_ebs/aws_util.go]: Did not notice this TODO last time :)

**Comment by swagiaal** [comment]
@thockin So it turns out the AWS EBS plugin was using the safe_format_and_mount script as well so I moved it to the mount package and now it shared between GCE PD and AWS EBS... WDYT ?

**Comment by markturansky** [comment]
@swagiaal we need to make sure these upstream PRs that were patched into OpenShift stay in sync downstream.

**Comment by swagiaal** [comment]
@thockin Does this still have your LGTM after expansion to AWS EBS ?

**Comment by swagiaal** [comment]
@thockin could you take another look at this ?

**Comment by saad-ali** [comment]
@swagiaal There have been a ton of changes since this was created, could you rebase?

**Comment by swagiaal** [comment]
@saad-ali rebased... thanks for taking a look!

File: pkg/util/mount/mount.go [hunk 14]
```diff
@@ -44,6 +49,69 @@ type MountPoint struct {
...
+	return err
+}
+
+// diskLooksUnformatted uses 'file' to see if the given disk is unformated
+func (mounter *SafeFormatAndMount) diskLooksUnformatted(disk string) (bool, error) {
+	args := []string{"-L", "--special-files", disk}
+	cmd := mounter.Runner.Command("file", args...)
+	dataOut, err := cmd.CombinedOutput()
+
+	// TODO (swagiaal): check if this disk has partitions and return false, and
```
Comment (saad-ali) [inline pkg/util/mount/mount.go]: Please open an issue to track this TODO and update this comment to include to issue #.

**Comment by saad-ali** [comment]
Thanks for the clean up here @swagiaal 
PR LGTM mod one minor nit

**Comment by swagiaal** [comment]
@saad-ali thanks for reviewing... comment addressed

**Comment by saad-ali** [comment]
Thanks @swagiaal 
LGTM
